[class-parse] Merge javadoc documentation

jonpryor opened the pull request:
Context: https://github.com/xamarin/java.interop/pull/623
Context: https://github.com/xamarin/java.interop/pull/623#issuecomment-666554407

DO NOT MERGE UNTIL AFTER PR #623 IS MERGED.

Update `class-parse --docspath=PATH` so that if `PATH` contains
`<javadoc/>` elements, as produced by `tools/java-source-utils` (PR #623),
then those `<javadoc/>` elements will be inserted into the generated
API description.

The intent is to eventually allow `generator` to emit the `<javadoc/>`
data as C# XML Documentation, allowing a pipeline of:

        java -jar java-source-tools path/to/android.jar --output-javadoc android.xml
        mono class-parse.exe --docspath=android.xml -o api.xml …
        mono generator.exe api.xml …

jpobst commented:
I wonder if we should try to do the "api documentation" part of this directly in `generator` instead of `class-parse`.

Doing it in `class-parse` means we have to read/write/persist it through all of the various `xml` files that our pipelines consists of:
```
api.xml.class-parse -> api.xml -> api.xml.fixed
```

I suspect the resulting `xml` is quite large.

On the other hand if it runs after we apply `metadata` it's probably going to make matching them harder.

Is the intention to run this on every CI build?  Once or for each API level?  How long does it take to run on `Mono.Android.dll`?

jonpryor commented:
@jpobst wrote:

> I wonder if we should try to do the "api documentation" part of this directly in generator instead of class-parse.

Probably!  Half the point here is to figure out how this should best be structured.  I've implemented it so that there's only "one" `api.xml` file, and -- with PR #685 -- `api.xml.fixed` contains the `<javadoc/>` elements.

Just because it "works" that way doesn't mean it's *correct* to do it that way.

> I suspect the resulting `xml` is quite large.

Indeed. For API-29 -- because my API-30 doesn't have any Javadoc comments! -- `java -jar java-source-tools path/to/android.jar --output-javadoc android.xml` emits a 38MB file.  After `generator` invocation, with PR #685, `api.xml.fixed` and `api.xml.adjusted` and `api.xml.adjusted.fixed` are 48MB in size.

> On the other hand if it runs after we apply metadata it's probably going to make matching them harder/impossible.

The way PR #684 works is by using `//@jni-signature` as the way to link things together.  That should be a fairly reliable way to match things, hopefully, except for the TODO in PR #623 regarding "what about unresolvable types?", which I forgot to take into consideration. :-/

> Is the intention to run this on every CI build?

Yes.

> Once or for each API level?

Ideally, we'd run `java-source-tool.jar` against only *one* API level, API-30, and reuse those docs for all API level builds.

This might not actually be possible because of parameter name changes between API levels, e.g. using the `android-javadocs.xml` from API-29 w/ API-21 will almost certainly change parameter names, so this requires more thought.

Additionally, my currently installed API-30 contains *no* Javadoc information.  (Perhaps a more recent API-30 will contain docs?)  Thus, for *now*, we'd need to use the API-29 docs when emitting API-30 files.

> How long does it take to run on Mono.Android.dll?

I'm not quite sure what you're asking here?  `java-source-tool.jar` invocations for API-29's `platforms/android-29/android-stubs-src.jar` takes ~90 seconds to run.  It's not *very* long, but it's certainly noticeable.

jpobst commented:
That is actually much faster and smaller than I was expecting.  I was expecting it to take ~5 minutes and generate ~500 MB files.  :) 

When we integrate it into the `Mono.Android.dll` build, let's flag it as something the CI opts into so that it doesn't run by default.  There are days where I build `Mono.Android.dll` locally 30+ times and it takes long enough already.

jonpryor commented:
> There are days where I build `Mono.Android.dll` locally 30+ times and it takes long enough already.

That's a statement that also confuses me.  The way I'm envisioning this -- subject to "meeting reality"! -- is that we'd generate an `android-javadocs.xml` file "once", a'la `bin/BuildDebug/api/api-29.xml.in` & co, which would be merged into `api.xml`. I don't see this as significantly altering the actual build time, but it's certainly something to investigate…

jonpryor commented:
Superseded by PR#687: We'll add a `generator --with-javadoc-xml` option, and have `generator` "merge" the documentation with the members at code emit time, instead of having `class-parse` do the merge.

Bonus: it means PR #685 is superseded, and it *doesn't* need to update the `api.xml` verification code in `src/Xamarin.Android.Tools.ApiXmlAdjuster`!
